lib/mutgen.c

```diff
@@ -25,66 +25,350 @@
+    }
+    if (!doubles_almost_equal(prob, 1.0, 1e-12) || min_prob < 0.0) {
+        ret = MSP_ERR_BAD_ROOT_PROBABILITIES;
+        goto out;
+    }
+    for (j = 0; j < self->alphabet_length; j++) {
+        prob = 0;
+        min_prob = 0;
+        mutation_row = self->mutation_matrix + j * self->alphabet_length;
+        for (k = 0; k < self->alphabet_length; k++) {
+            prob += mutation_row[k];
+            min_prob = GSL_MIN(min_prob, mutation_row[k]);
+        }
+        if (!doubles_almost_equal(prob, 1.0, 1e-12) || min_prob < 0.0) {
+            ret = MSP_ERR_BAD_MUTATION_MATRIX;
+            goto out;
+        }
+    }
+
+out:
+    return ret;
+}
+
+int MSP_WARN_UNUSED
+mutation_model_alloc(mutation_model_t *self, int alphabet_length,
+        const char *alphabet, double *root_distribution, double *mutation_matrix)
+{
+    int ret = 0;
+    int j, k, jk;
+    char *buff = NULL;
+
+    memset(self, 0, sizeof(mutation_model_t));
+    if (alphabet_length < 2) {
+        ret = MSP_ERR_INSUFFICIENT_ALLELES;
+        goto out;
+    }
+
+    buff = malloc((size_t) alphabet_length * sizeof(*self->alphabet));
+    self->root_distribution = malloc((size_t) alphabet_length * sizeof(*self->root_distribution));
+    self->mutation_matrix = malloc((size_t) (alphabet_length * alphabet_length)
+                                            * sizeof(*self->mutation_matrix));
+    if (self->root_distribution == NULL
+            || self->mutation_matrix == NULL) {
+        ret = MSP_ERR_NO_MEMORY;
+        goto out;
+    }
+    self->alphabet_length = alphabet_length;
+    for (j = 0; j < alphabet_length; j++) {
+        buff[j] = alphabet[j];
+        self->root_distribution[j] = root_distribution[j];
+        for (k = 0; k < alphabet_length; k++) {
+            jk = j * alphabet_length + k;
+            self->root_distribution[jk] = mutation_matrix[jk];
+        }
+    }
+    self->alphabet = buff;
+    ret = mutation_model_check_validity(self);
+out:
+    return ret;
+}
+
+int
+mutation_model_free(mutation_model_t *self)
+{
+    if (self->alphabet != NULL) {
+        free(&self->alphabet);
+    }
+    msp_safe_free(self->root_distribution);
+    msp_safe_free(self->mutation_matrix);
+    return 0;
+}
+
+int
+mutation_model_get_alphabet_length(mutation_model_t *self)
+{
+    return self->alphabet_length;
+}
+
 static int
-cmp_site(const void *a, const void *b) {
-    const tsk_site_t *ia = (const tsk_site_t *) a;
-    const tsk_site_t *ib = (const tsk_site_t *) b;
-    return (ia->position > ib->position) - (ia->position < ib->position);
+mutation_model_allele_index(mutation_model_t *self, const char *allele, size_t length)
+{
+    /* returning -1 means "not found" */
+    int ret = 0;
+    if (length != 1) {
+        /* only single-character alphabets for now */
+        ret = -1;
+        goto out;
+    }
+    while (ret < self->alphabet_length && self->alphabet[ret] != *allele) {
```

Comment:
I don't think this works. Here's how we do the same thing in [tskit](https://github.com/tskit-dev/tskit/blob/master/c/tskit/genotypes.c#L472)